name.abuchen.portfolio.ui/src/name/abuchen/portfolio/ui/util/chart/ChartToolsManager.java

```diff
@@ -0,0 +1,523 @@
+package name.abuchen.portfolio.ui.util.chart;
+
+import java.text.DecimalFormat;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+
+import org.eclipse.jface.action.IAction;
+import org.eclipse.jface.action.IMenuManager;
+import org.eclipse.jface.action.ToolBarManager;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.events.PaintEvent;
+import org.eclipse.swt.graphics.Color;
+import org.eclipse.swt.graphics.Point;
+import org.eclipse.swt.graphics.Rectangle;
+import org.eclipse.swt.widgets.Event;
+import org.swtchart.IAxis;
+
+import name.abuchen.portfolio.money.Values;
+import name.abuchen.portfolio.ui.Images;
+import name.abuchen.portfolio.ui.Messages;
+import name.abuchen.portfolio.ui.util.Colors;
+import name.abuchen.portfolio.ui.util.SimpleAction;
+import name.abuchen.portfolio.util.Pair;
+
+public class ChartToolsManager
+{
+    private abstract static class ChartToolBase
+    {
+
+        void onMouseDown(Event e)
+        {
+        }
+
+        void onMouseMove(Event e)
+        {
+        }
+
+        void onMouseUp(Event e)
+        {
+        }
+
+        void paintControl(PaintEvent e)
+        {
+        }
+
+    }
+
+    public enum ChartTools
+    {
+        NONE("", c -> null, null, null), // //$NON-NLS-1$
+        CROSSHAIR(Messages.BtnLabelCrosshair, p -> new CrosshairTool(p.getLeft(), p.getRight()), Images.CROSSHAIR_ON, 
+                        Images.CROSSHAIR_OFF), //
+        MEASUREMENT(Messages.BtnLabelMeasureDistance, p -> new MeasurementTool(p.getLeft(), p.getRight()),
+                        Images.MEASUREMENT_ON, Images.MEASUREMENT_OFF);
+
+        private String actionText;
+        private Function<Pair<TimelineChart, Color>, ChartToolBase> createToolFactory;
```

Comment:
Let's not introduce more usages for the generic ```Pair``` class.
You can use Java's ```BiFunction``` here:

```
private BiFunction<TimelineChart, Color, ChartToolBase> factory;
```

Reply:
Ah yes, don't know why I didn't found the ``BiFunction``. It's much better.